### DISCUSSION TITLE
Zephyr + ExecuTorch + Alif E8: ML inference in 19ms on Ethos-U55

### DISCUSSION THREAD

**Discussion by psiddh** [zephyrproject-rtos/zephyr#110220] [Show and tell]
Wanted to share a project where we got [ExecuTorch](https://docs.pytorch.org/executorch/main/index.html) (PyTorch's solution for on-device AI) running as a native Zephyr module on the Alif Ensemble E8 DevKit.

The setup: a quantized MobileNetV2 (ImageNet classifier), fully delegated to the Ethos-U55 NPU — all 110 ops on the NPU, nothing on CPU. Inference completes in about 19ms.

<img width="4000" height="3000" alt="20260601_002648" src="https://github.com/user-attachments/assets/edd001ed-8b47-42bb-a880-165e2681fe1b" />

<img width="1713" height="1035" alt="image" src="https://github.com/user-attachments/assets/17379780-ad16-4697-bfda-ddce4afad833" />

## How it works

1. Export a PyTorch model with INT8 quantization and Ethos-U delegation (Vela compiler handles the NPU command stream)
2. Add ExecuTorch to your Zephyr workspace as a module via `west` — it has a proper `module.yml`, Kconfig, and CMake integration
3. Build and flash like any Zephyr app:

```bash
west build -b alif_e8_dk/ae822fa0e5597xx0/rtss-he -p
west flash
```

## Numbers

| Metric | Value |
|---|---|
| Inference latency | ~19ms (224x224 RGB input) |
| NPU ops | 110/110 delegated to Ethos-U55 |
| Model size | ~3.5 MB (PTE format) |
| Memory | Model in MRAM, buffers in 4.5 MB HP SRAM |

## What we learned

- The MRAM flash address needs to be `0x80008000` instead of the default `0x80200000`, otherwise the model doesn't fit in the remaining space. Small thing but cost us some debugging time.
- FVP simulation is significantly slower than real hardware, having an actual E8 board makes development much more practical.
- The demo currently uses untrained weights (zero predictions) because pretrained MV2 inflates to ~14MB after Vela compilation, exceeding the E8's 5.5MB MRAM. The speed and NPU delegation are real — working on loading from external OSPI flash for actual predictions.

## Links

- Full step-by-step tutorial (official PyTorch docs): https://docs.pytorch.org/executorch/main/zephyr_alif_tutorial.html
- ExecuTorch repo: https://github.com/pytorch/executorch

If anyone has an E8 board and tries this out, happy to help with any issues.

**Comment by digantdesai** [comment]
Love it. Thanks @psiddh !
